src/test/java/io/github/christophermanahan/carnitas/HTTPServerTest.java

```diff
@@ -3,163 +3,180 @@
-        String request = "GET /simple_get HTTP/1.1";
-        Acceptor acceptor = new TestAcceptor(List.of(new SendException(request)));
+    void itWillLogAMessageIfListenFails() {
+        String message = "Failed!";
+        Listener listener = () -> {
+            throw new RuntimeException(message);
+        };
 
-        new HTTPServer(acceptor, logger).run();
+        new HTTPServer(parser, handler, logger).start(listener, new Once());
 
-        assertEquals(ErrorMessages.SEND_TO_CONNECTION, logger.logged());
+        assertEquals(message, logger.logged());
     }
 
     @Test
-    void logsExceptionIfCloseFails() {
-        String request = "GET /simple_get HTTP/1.1";
-        Acceptor acceptor = new TestAcceptor(List.of(new CloseException(request)));
-
-        new HTTPServer(acceptor, logger).run();
-
-        assertEquals(ErrorMessages.CLOSE_CONNECTION, logger.logged());
+    void itWillLogAMessageIfSendFails() {
+        String message = "Failed!";
+        Connection connection = new Connection() {
+            private int index = -1;
+
+            public void send(HTTPResponse response) {
+                throw new RuntimeException(message);
+            }
+
+            public void close() {
+            }
+
+            public Optional<Character> read() {
+                index++;
+                return Optional.of(
+                  List.of(HTTPResponse.CRLF.split("")).get(index).charAt(0)
+                );
+            }
+        };
+        Listener listener = () -> connection;
+
+        new HTTPServer(parser, handler, logger).start(listener, new Once());
+
+        assertEquals(message, logger.logged());
     }
 
-    private class TestLogger implements Logger {
-        private final StringBuilder log;
-
-        TestLogger() {
-            this.log = new StringBuilder();
-        }
+    private class TestListener implements Listener {
+        private final Iterator<ReadableConnection> connections;
 
-        String logged() {
-            return log.toString();
+        TestListener(List<ReadableConnection> connections) {
+            this.connections = connections.iterator();
         }
 
-        public void log(String message) {
-            log.append(message);
+        public Connection listen() {
+            return connections.next();
         }
     }
 
-    private class TestAcceptor implements Acceptor {
-        final Iterator<TestConnection> connections;
+    private class ReadableConnection implements Connection {
+        private final String request;
+        private int index;
+        HTTPResponse response;
 
-        TestAcceptor(List<TestConnection> connections) {
-            this.connections = connections.iterator();
+        ReadableConnection(String request) {
+            this.request = request;
+            this.index = -1;
         }
 
-        public TestConnection accept() {
-            return connections.next();
+        public Optional<Character> read() {
+            index++;
```

Comment:
Is there any way for us to avoid using indexes like this?

Reply:
Yes, switched over to an iterator!